test/lit_tests/diffrules/stablehlo/while3.mlir

```diff
@@ -0,0 +1,98 @@
+// CHECK-NEXT:    %c = stablehlo.constant dense<3> : tensor<i64>
+// CHECK-NEXT:    %c_0 = stablehlo.constant dense<1> : tensor<i64>
+// CHECK-NEXT:    %c_1 = stablehlo.constant dense<0> : tensor<i64>
+// CHECK-NEXT:    %cst = arith.constant dense<0.000000e+00> : tensor<f64>
+// CHECK-NEXT:    %0 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%0, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %1 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%1, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %2 = "enzyme.init"() : () -> !enzyme.Cache<tensor<f64>>
+// CHECK-NEXT:    %3 = "enzyme.init"() : () -> !enzyme.Cache<tensor<f64>>
+// CHECK-NEXT:    %4 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%4, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %5 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%5, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %6:2 = stablehlo.while(%iterArg = %c_1, %iterArg_2 = %arg0) : tensor<i64>, tensor<f64>
+// CHECK-NEXT:     cond {
+// CHECK-NEXT:      %14 = stablehlo.compare  LT, %iterArg, %c : (tensor<i64>, tensor<i64>) -> tensor<i1>
+// CHECK-NEXT:      stablehlo.return %14 : tensor<i1>
+// CHECK-NEXT:    } do {
+// CHECK-NEXT:      "enzyme.push"(%3, %iterArg_2) : (!enzyme.Cache<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      "enzyme.push"(%2, %iterArg_2) : (!enzyme.Cache<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      %14 = stablehlo.multiply %iterArg_2, %iterArg_2 : tensor<f64>
+// CHECK-NEXT:      %15 = stablehlo.add %iterArg, %c_0 : tensor<i64>
+// CHECK-NEXT:      stablehlo.return %15, %14 : tensor<i64>, tensor<f64>
+// CHECK-NEXT:    }
+// CHECK-NEXT:    %7 = "enzyme.get"(%5) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:    %8 = arith.addf %7, %arg1 : tensor<f64>
+// CHECK-NEXT:    "enzyme.set"(%5, %8) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %9 = "enzyme.get"(%5) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:    "enzyme.set"(%5, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %10:2 = stablehlo.while(%iterArg = %c_1, %iterArg_2 = %9) : tensor<i64>, tensor<f64>
+// CHECK-NEXT:     cond {
+// CHECK-NEXT:      %14 = stablehlo.compare  LT, %iterArg, %c : (tensor<i64>, tensor<i64>) -> tensor<i1>
+// CHECK-NEXT:      stablehlo.return %14 : tensor<i1>
+// CHECK-NEXT:    } do {
+// CHECK-NEXT:      %14 = stablehlo.add %iterArg, %c_0 : tensor<i64>
+// CHECK-NEXT:      "enzyme.set"(%4, %iterArg_2) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      %15 = "enzyme.get"(%4) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
```

Comment:
we should make sure that remove-enzyme-ops can deal with this %4 case without extra work.

%4 is defined right before this in a dominating way allowing %15 to be replaced with %iterArg_2.

Then now that only set's are called on it, we should be able to remove the whole thing.

No while special case handling required here